## Тема “Обучение с учителем”

### Задание 1
#### Импортируйте библиотеки pandas и numpy.
#### Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn.

In [50]:
# код для подавления предупреждений в процессе выполнения дальнейшего кода
import warnings
warnings.filterwarnings('ignore')

In [51]:
import pandas as pd
import numpy as np

#### Библиотека `scikit-learn` с версии 1.2 убрала функцию `load_boston`, поскольку набор данных о ценах на жильё в Бостоне имеет этические проблемы. Версия 1.2 была выпущена в декабре 2021 года. Урок, записанный мужским голосом, от 24.06.2018 г. - немного устарело.
#### Заменим использование `load_boston` на один из доступных наборов данных, например, `fetch_california_housing` для данных о жилье в Калифорнии или `fetch_openml` для набора данных о ценах на жилье в Ames.

In [52]:
from sklearn.datasets import fetch_california_housing

In [53]:
fch = fetch_california_housing()

In [54]:
# содержимое датасета
fch.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'feature_names', 'DESCR'])

In [55]:
data = fch.data

In [56]:
data.shape

(20640, 8)

In [57]:
data

array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
          37.88      , -122.23      ],
       [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
          37.86      , -122.22      ],
       [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
          37.85      , -122.24      ],
       ...,
       [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
          39.43      , -121.22      ],
       [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
          39.43      , -121.32      ],
       [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
          39.37      , -121.24      ]])

In [58]:
target = fch.target # нужно предсказать эту величину

In [59]:
feature_names = fch["feature_names"]
# feature_names = fch.feature_names
feature_names # названия признаков

['MedInc',
 'HouseAge',
 'AveRooms',
 'AveBedrms',
 'Population',
 'AveOccup',
 'Latitude',
 'Longitude']

In [60]:
# print(fch["DESCR"]) # описание датасета

# информация о датасете, признаках, а также об авторах
for line in fch.DESCR.split('\n'):
    print(line)

.. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

    :Number of Instances: 20640

    :Number of Attributes: 8 numeric, predictive attributes and the target

    :Attribute Information:
        - MedInc        median income in block group
        - HouseAge      median house age in block group
        - AveRooms      average number of rooms per household
        - AveBedrms     average number of bedrooms per household
        - Population    block group population
        - AveOccup      average number of household members
        - Latitude      block group latitude
        - Longitude     block group longitude

    :Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median house value for California districts,
expressed in hundreds of thousands of dollars ($100,000).

This dataset was derived

#### Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random state должен быть равен 42.

In [61]:
X = pd.DataFrame(data, columns=feature_names)
X.head() # признаки недвижимости

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [62]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Latitude    20640 non-null  float64
 7   Longitude   20640 non-null  float64
dtypes: float64(8)
memory usage: 1.3 MB


In [63]:
y = pd.DataFrame(target, columns=["price"])
y.head() # целевые значения

price
0  4.526
1  3.585
2  3.521
3  3.413
4  3.422

In [64]:
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   price   20640 non-null  float64
dtypes: float64(1)
memory usage: 161.4 KB


#### Разбиение выборки на тренировочную и тестовую

In [65]:
from sklearn.model_selection import train_test_split

In [66]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
# размер тестовой выборки test_size равен 30%, random_state=42 - повторимость случайного выбора

#### Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

In [67]:
from sklearn.linear_model import LinearRegression

In [68]:
lr = LinearRegression()

In [69]:
lr.fit(X_train, y_train)
print(lr.coef_)

[[ 4.45822565e-01  9.68186799e-03 -1.22095112e-01  7.78599557e-01
  -7.75740400e-07 -3.37002667e-03 -4.18536747e-01 -4.33687976e-01]]


In [70]:
y_pred = lr.predict(X_test)
y_pred.shape
# получить предсказанные значения на объектах X_test с помощью метода .predict:

(6192, 1)

#### Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

In [71]:
# чтобы сопоставить реальные значения с предсказанными
# метод .flatten - чтобы перевести двумерный массив y_pred в одномерный
check_test = pd.DataFrame({
    "y_test": y_test["price"],
    "y_pred": y_pred.flatten(),
})

check_test.head(20)

y_test    y_pred
20046  0.47700  0.726049
3024   0.45800  1.767434
15663  5.00001  2.710922
20484  2.18600  2.835147
9814   2.78000  2.606958
13311  1.58700  2.010739
7113   1.98200  2.640674
7668   1.57500  2.167062
18246  3.40000  2.740121
5723   4.46600  3.903615
20069  1.23200  0.944419
6835   2.53900  1.901237
11351  2.15100  1.763534
20267  2.20500  2.252804
7097   2.19800  2.537990
6298   1.36200  1.922175
696    1.78400  2.387974
19607  1.87500  2.012288
14173  1.39800  2.224112
19638  1.37500  1.126040

##### _Таким образом, далеко не все цены предсказаны точно. В таблице y_test и y_pred разнятся от 0.047 (20267: 2.20500 - 2.252804) до 0.66 (18246: 3.40000 - 2.740121) - считал вручную сам._
##### Посчитаем также R² (коэффициент детерминации) - он показывает, какая доля изменчивости целевой переменной может быть объяснена моделью. Значение R² варьируется от -∞ до 1, где 1 означает идеальное предсказание.

In [72]:
# Считаем R2 (коэффициент детерминации) для модели линейной регрессии
from sklearn.metrics import r2_score
r2_lr = r2_score(y_test, y_pred)
print(f"R2 для модели линейной регрессии = {r2_lr}")

R2 для модели линейной регрессии = 0.5957702326061665


### Задание 2
#### Создайте модель под названием model с помощью класса RandomForestRegressor из модуля sklearn.ensemble.
#### Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42.
#### Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
#### Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
#### Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [73]:
from sklearn.ensemble import RandomForestRegressor # gb.ru/lessons/443634#s2489
#from sklearn.model_selection import GridSearchCV

In [74]:
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [75]:
# Сделайте предсказание на тестовых данных
y_pred_rf = model.predict(X_test)
y_pred_rf.shape

(6192,)

In [76]:
# Расчет R2 для Random Forest
r2_rf = r2_score(y_test, y_pred_rf)
print(f'R2 score для Random Forest = {r2_rf}')

R2 score для Random Forest = 0.7945055702215934


In [77]:
# Сравниваем результаты
if r2_rf > r2_lr:
    print("Модель RandomForestRegressor ближе к 1 и работает лучше.")
else:
    print("Модель линейной регрессии ближе к 1 и работает лучше.")

Модель RandomForestRegressor ближе к 1 и работает лучше.


### *Задание 3
#### Вызовите документацию для класса RandomForestRegressor, найдите информацию об атрибуте feature_importances_.
#### С помощью этого атрибута найдите сумму всех показателей важности, установите, какие два признака показывают наибольшую важность.

In [78]:
RandomForestRegressor?
# help(RandomForestRegressor)

Init signature:
RandomForestRegressor(
    n_estimators=100,
    *,
    criterion='squared_error',
    max_depth=None,
    min_samples_split=2,
    min_samples_leaf=1,
    min_weight_fraction_leaf=0.0,
    max_features=1.0,
    max_leaf_nodes=None,
    min_impurity_decrease=0.0,
    bootstrap=True,
    oob_score=False,
    n_jobs=None,
    random_state=None,
    verbose=0,
    warm_start=False,
    ccp_alpha=0.0,
    max_samples=None,
)
Docstring:     
A random forest regressor.

A random forest is a meta estimator that fits a number of classifying
decision trees on various sub-samples of the dataset and uses averaging
to improve the predictive accuracy and control over-fitting.
The sub-sample size is controlled with the `max_samples` parameter if
`bootstrap=True` (default), otherwise the whole dataset is used to build
each tree.

Read more in the :ref:`User Guide <forest>`.

Parameters
----------
n_estimators : int, default=100
    The number of trees in the forest.

    .. versioncha

##### feature_importances_ : ndarray of shape (n_features,)
#####     The impurity-based feature importances.
#####     The higher, the more important the feature.
#####     The importance of a feature is computed as the (normalized)
#####     total reduction of the criterion brought by that feature.  It is also
#####     known as the Gini importance.

#####     Warning: impurity-based feature importances can be misleading for
#####     high cardinality features (many unique values). See
#####     :func:`sklearn.inspection.permutation_importance` as an alternative.

In [79]:
# сумма показателей важности
sum_importance = np.sum(model.feature_importances_)
sum_importance # всегда равна 1.0

1.0

In [80]:
# Создание таблицы с показателями важности
importance_df = pd.DataFrame({
    'Показатель': model.feature_importances_,
    'Имя': X.columns,
})

In [81]:
# Сортировка таблицы по показателю важности в порядке убывания
importance_df = importance_df.sort_values(by='Показатель', ascending=False).reset_index(drop=True)
importance_df

Показатель         Имя
0    0.559269      MedInc
1    0.139125    AveOccup
2    0.082233   Longitude
3    0.081538    Latitude
4    0.050431    HouseAge
5    0.039843    AveRooms
6    0.024307  Population
7    0.023255   AveBedrms

In [82]:
top_features = importance_df.nlargest(2, 'Показатель')
top_features

Показатель       Имя
0    0.559269    MedInc
1    0.139125  AveOccup

### *Задание 4 - *выполнил НЕ до конца, наполовину*
#### В этом задании мы будем работать с датасетом, с которым мы уже знакомы по домашнему заданию по библиотеке Matplotlib, это датасет Credit Card Fraud Detection. Для этого датасета мы будем решать задачу классификации - будем определять, какие из транзакции по кредитной карте являются мошенническими. Данный датасет сильно несбалансирован (так как случаи мошенничества относительно редки), так что применение метрики accuracy не принесет пользы и не поможет выбрать лучшую модель. Мы будем вычислять AUC, то есть площадь под кривой ROC.
    
#### Импортируйте из соответствующих модулей RandomForestClassifier, GridSearchCV и train_test_split.

#### Загрузите датасет creditcard.csv и создайте датафрейм df.

#### С помощью метода value_counts с аргументом normalize=True убедитесь в том, что выборка несбалансирована. Используя метод info, проверьте, все ли столбцы содержат числовые данные и нет ли в них пропусков. Примените следующую настройку, чтобы можно было просматривать все столбцы датафрейма:
#### pd.options.display.max_columns = 100.
#### Просмотрите первые 10 строк датафрейма df.
#### Создайте датафрейм X из датафрейма df, исключив столбец Class.
#### Создайте объект Series под названием y из столбца Class.
#### Разбейте X и y на тренировочный и тестовый наборы данных при помощи функции train_test_split, используя аргументы: test_size=0.3, random_state=100, stratify=y.
#### У вас должны получиться объекты X_train, X_test, y_train и y_test.
#### Просмотрите информацию о их форме.
#### Для поиска по сетке параметров задайте такие параметры:
#### parameters = [{'n_estimators': [10, 15],
#### 'max_features': np.arange(3, 5),
#### 'max_depth': np.arange(4, 7)}]
#### Создайте модель GridSearchCV со следующими аргументами:
#### estimator=RandomForestClassifier(random_state=100),
#### param_grid=parameters,
#### scoring='roc_auc',
#### cv=3.
#### Обучите модель на тренировочном наборе данных (может занять несколько минут).
#### Просмотрите параметры лучшей модели с помощью атрибута best_params_.
#### Предскажите вероятности классов с помощью полученной модели и метода predict_proba.
#### Из полученного результата (массив Numpy) выберите столбец с индексом 1 (вероятность класса 1) и запишите в массив y_pred_proba. Из модуля sklearn.metrics импортируйте метрику roc_auc_score.
#### Вычислите AUC на тестовых данных и сравните с результатом,полученным на тренировочных данных, используя в качестве аргументов массивы y_test и y_pred_proba.

In [83]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score # для оценки производительности модели классификации, особенно с несбалансированными классами

In [84]:
df = pd.read_csv('creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10       V11       V12       V13       V14  \
0  0.098698  0.363787  0.090794 -0.551600 -0.617801 -0.991390 -0.311169   
1  0.085102 -0.255425 -0.166974  1.612727  1.065235  0.489095 -0.143772   
2  0.247676 -1.514654  0.207643  0.624501  0.066084  0.717293 -0.165946   
3  0.377436 -1.387024 -0.054952 -0.226487  0.178228  0.507757 -0.287924   
4 -0.270533  0.817739  0.753074 -0.822843  0.538196  1.345852 -1.119670   

        V15       V16       V17       V18       V19       V20       V21  \
0  1.468177 -0.470401  0.207971  0.025791  0.403993  0.251412 -0.018307   
1  0.635558  0.463917 -0.114805 -0.183361 -0.145783 -0.069083 -0.225775   
2  2.345865 -2.890083  1.109969 -0.121359 -2.261857  0.524980  0.247998   
3 -0.631418 -1.059647 -0.684093  1.965775 -1.232622 -0.208038 -0.108300   
4  0.175121 -0.451449 -0.237033 -0.038195  0.803487  0.408542 -0.009431   

        V22       V23       V24       V25       V26       V27       V28  \
0  0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1 -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2  0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3  0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4  0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   

   Amount  Class  
0  149.62      0  
1    2.69      0  
2  378.66      0  
3  123.50      0  
4   69.99      0

In [85]:
df.info() # у всех столбцов тип float64 или int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [86]:
pd.options.display.max_columns = 100
df.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9       V10       V11       V12       V13       V14  \
0  0.098698  0.363787  0.090794 -0.551600 -0.617801 -0.991390 -0.311169   
1  0.085102 -0.255425 -0.166974  1.612727  1.065235  0.489095 -0.143772   
2  0.247676 -1.514654  0.207643  0.624501  0.066084  0.717293 -0.165946   
3  0.377436 -1.387024 -0.054952 -0.226487  0.178228  0.507757 -0.287924   
4 -0.270533  0.817739  0.753074 -0.822843  0.538196  1.345852 -1.119670   
5  0.260314 -0.568671 -0.371407  1.341262  0.359894 -0.358091 -0.137134   
6  0.081213  0.464960 -0.099254 -1.416907 -0.153826 -0.751063  0.167372   
7 -3.807864  0.615375  1.249376 -0.619468  0.291474  1.757964 -1.323865   
8  0.851084 -0.392048 -0.410430 -0.705117 -0.110452 -0.286254  0.074355   
9  0.069539 -0.736727 -0.366846  1.017614  0.836390  1.006844 -0.443523   

        V15       V16       V17       V18       V19       V20       V21  \
0  1.468177 -0.470401  0.207971  0.025791  0.403993  0.251412 -0.018307   
1  0.635558  0.463917 -0.114805 -0.183361 -0.145783 -0.069083 -0.225775   
2  2.345865 -2.890083  1.109969 -0.121359 -2.261857  0.524980  0.247998   
3 -0.631418 -1.059647 -0.684093  1.965775 -1.232622 -0.208038 -0.108300   
4  0.175121 -0.451449 -0.237033 -0.038195  0.803487  0.408542 -0.009431   
5  0.517617  0.401726 -0.058133  0.068653 -0.033194  0.084968 -0.208254   
6  0.050144 -0.443587  0.002821 -0.611987 -0.045575 -0.219633 -0.167716   
7  0.686133 -0.076127 -1.222127 -0.358222  0.324505 -0.156742  1.943465   
8 -0.328783 -0.210077 -0.499768  0.118765  0.570328  0.052736 -0.073425   
9  0.150219  0.739453 -0.540980  0.476677  0.451773  0.203711 -0.246914   

        V22       V23       V24       V25       V26       V27       V28  \
0  0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1 -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2  0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3  0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4  0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   
5 -0.559825 -0.026398 -0.371427 -0.232794  0.105915  0.253844  0.081080   
6 -0.270710 -0.154104 -0.780055  0.750137 -0.257237  0.034507  0.005168   
7 -1.015455  0.057504 -0.649709 -0.415267 -0.051634 -1.206921 -1.085339   
8 -0.268092 -0.204233  1.011592  0.373205 -0.384157  0.011747  0.142404   
9 -0.633753 -0.120794 -0.385050 -0.069733  0.094199  0.246219  0.083076   

   Amount  Class  
0  149.62      0  
1    2.69      0  
2  378.66      0  
3  123.50      0  
4   69.99      0  
5    3.67      0  
6    4.99      0  
7   40.80      0  
8   93.20      0  
9    3.68      0

In [87]:
print(df['Class'].value_counts(normalize=True)) # проверка несбалансированности выборки

Class
0    0.998273
1    0.001727
Name: proportion, dtype: float64


##### Почти все транзакции (99,8%) являются нормальными. Только 0.2% транзакций являются мошенническими.
##### Если один класс составляет больше, чем 70-80% от общего числа записей, то имеем дело с несбалансированной выборкой. В задачах классификации, где один из классов составляет менее 10% от общего числа, выборка считается сильно несбалансированной.

In [88]:
X = df.drop(columns=['Class'])
y = df['Class'] # создание датафрейма X и объекта y

In [89]:
# разделение на тренировочную и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=100, stratify=y)
print(f'Shape of X_train: {X_train.shape}, X_test: {X_test.shape}, y_train: {y_train.shape}, y_test: {y_test.shape}')

Shape of X_train: (199364, 30), X_test: (85443, 30), y_train: (199364,), y_test: (85443,)
